File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskKey.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask;

import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.view.View;

import com.google.auto.value.AutoValue;
import com.zhuinden.simplestack.navigator.ViewChangeHandler;
import com.zhuinden.simplestack.navigator.changehandlers.SegueViewChangeHandler;
import com.zhuinden.simplestackdemoexamplemvp.R;
import com.zhuinden.simplestackdemoexamplemvp.application.Key;

/**
 * Created by Zhuinden on 2017.01.25..
 */

@AutoValue
public abstract class AddOrEditTaskKey
        implements Key {
    public abstract Key parent();

    public abstract String taskId();

    public static Parcelable create(Key parent) {
        return createWithTaskId(parent, "");
    }

    public static Parcelable createWithTaskId(Key parent, String taskId) {
        return new AutoValue_AddOrEditTaskKey(R.layout.path_addoredittask, parent, taskId);
    }

    @NonNull
    @Override
    public ViewChangeHandler viewChangeHandler() {
        return new SegueViewChangeHandler();
    }

    @Override
    public int menu() {
        return R.menu.empty_menu;
    }

    @Override
    public boolean isFabVisible() {
        return true;
    }

    @Override
    public int navigationViewId() {
        return 0;
    }

    @Override
    public boolean shouldShowUp() {
        return true;
    }

    @Override
    public View.OnClickListener fabClickListener(View view) {
        return v -> {
            AddOrEditTaskView addOrEditTaskCoordinator = (AddOrEditTaskView)view;
            addOrEditTaskCoordinator.saveTask();
            addOrEditTaskCoordinator.navigateBack();
        };
    }

    @Override
    public int fabDrawableIcon() {
        return R.drawable.ic_done;
    }
}

File: simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/addoredittask/AddOrEditTaskKey.kt
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask

import android.view.View
import com.zhuinden.simplestack.ScopedServices
import com.zhuinden.simplestack.navigator.ViewChangeHandler
import com.zhuinden.simplestack.navigator.changehandlers.SegueViewChangeHandler
import com.zhuinden.simplestackdemoexamplemvp.R
import com.zhuinden.simplestackdemoexamplemvp.application.Injector
import com.zhuinden.simplestackdemoexamplemvp.application.Key
import com.zhuinden.simplestackdemoexamplemvp.util.scoping.ServiceProvider
import kotlinx.android.parcel.Parcelize

sealed class AddOrEditTaskKey(val parent: Key, val taskId: String = "") : Key, ServiceProvider.HasServices {
    override fun getScopeTag() = "AddOrEditTask"

    override fun bindServices(serviceBinder: ScopedServices.ServiceBinder) {
        serviceBinder.add(AddOrEditTaskView.CONTROLLER_TAG, Injector.get().addOrEditTaskPresenter())
    }

    @Parcelize
    data class AddTaskKey(val parentKey: Key): AddOrEditTaskKey(parentKey)

    @Parcelize
    data class EditTaskKey(val parentKey: Key, val taskID: String): AddOrEditTaskKey(parentKey, taskID)

    override fun layout(): Int = R.layout.path_addoredittask

    override val isFabVisible: Boolean
        get() = true

    override fun viewChangeHandler(): ViewChangeHandler = SegueViewChangeHandler()

    override fun menu(): Int = R.menu.empty_menu

    override fun navigationViewId(): Int = 0

    override fun shouldShowUp(): Boolean = true

    override fun fabClickListener(view: View): View.OnClickListener =
        View.OnClickListener { v ->
            val addOrEditTaskView = view as AddOrEditTaskView
            addOrEditTaskView.fabClicked()
        }

    override fun fabDrawableIcon(): Int = R.drawable.ic_done
}
